lib/pbench/server/db/models/tracker.py

```diff
@@ -0,0 +1,299 @@
+    EXPIRED = ("Expired", False)
+
+    def __init__(self, friendly, mutating):
+        """
+        __init__ Extended ENUM constructor in order to add a value
+        to record whether each state is a "busy" state where some
+        component is mutating the dataset in some way, and a mixed
+        case "friendly" name for the state. "Mutating" states are
+        expected to be transient as the component should complete
+        the mutation, and should usually have "-ing" endings.
+
+        Args:
+            name (string): Friendly name for the state
+            mutating (boolean): True if a component is mutating dataset
+        """
+        self.friendly = friendly
+        self.mutating = mutating
+
+    def __str__(self):
+        "Return the state's friendly name"
+        return self.friendly
+
+
+class Dataset(Database.Base):
+    """ Identify a Pbench dataset (tarball plus metadata)
+
+    Columns:
+        id          Generated unique ID of table row
+        user        User associated with the dataset
+        controller  Name of controller node
+        name        Base name of dataset (tarball)
+        md5         The dataset MD5 hash (Elasticsearch ID)
+        created     Date the dataset was PUT to server
+        state       The current state of the dataset
+        transition  The timestamp of the last state transition
+
+    TODO: This could be expanded/generalized to support arbitrary user
+    metadata such as search keywords. We need to define some basic
+    parameters and figure out what goes here and what would be more
+    useful in the Elasticsearch DB.
+    """
+
+    __tablename__ = "dataset"
+
+    transitions = {
+        States.UPLOADING: [States.UPLOADED],
+        States.UPLOADED: [States.ARCHIVING],
+        States.ARCHIVING: [States.ARCHIVED],
+        States.ARCHIVED: [States.UNPACKING],
+        States.UNPACKING: [States.UNPACKED],
+        States.UNPACKED: [States.INDEXING],
+        States.INDEXING: [States.INDEXED],
+        States.INDEXED: [States.REINDEXING, States.EXPIRING],
+        States.REINDEXING: [States.INDEXED],
+        States.EXPIRING: [States.EXPIRED, States.INDEXED],
+    }
+
+    DEFAULT_USERNAME = "public"
+
+    id = Column(Integer, primary_key=True, autoincrement=True)
+    user = Column(String(255), unique=False, nullable=False)
+    controller = Column(String(255), unique=False, nullable=False)
+    name = Column(String(255), unique=False, nullable=False)
+    md5 = Column(String(255), unique=False, nullable=True)
+    created = Column(DateTime, nullable=False)
+    state = Column(Enum(States), unique=False, nullable=False)
+    transition = Column(DateTime, nullable=False)
+
+    def __init__(self, **kwargs):
+        """
+        Construct a new Dataset object; generally this only makes sense when
+        taking in a new dataset via PUT, and the state of the dataset defaults
+        to NEW.
+
+        However, we may also use this to "import" previous datasets pushed to
```

Comment:
Mostly, I wanted to test it easily; and the simplest path was to re-index the existing datasets on my test server. With shim wired to "import" I'll be able to push a new dataset with `pbench-move-results` ... then I'll need to figure out how to drive the `PUT` API to do it right... although of course we still don't have a real user.

Then again, the shim, like PUT, assumes there's no existing dataset record; it doesn't really "import" at all.

In any case, I wouldn't want to code a separate DB Model for "import" (or, even worse, do direct SQL bypassing the model) and, despite this little note in the docstring, there's no extra code involved in `__init__`. The real difference would be that we don't need an "attach or create" semantic since in every case we either know it doesn't exist yet (shim, PUT) or expect that it _does_ exist already. (And I can't remember whether I already pushed a commit with this change, but the current code has `attach` which fails if it can't find the dataset, and a new `attach_or_create` that just handles the `DatasetNotFound` exception and creates one... which _could_ easily be moved out of the `Dataset` class and done anywhere.)